crates/ty_python_semantic/resources/mdtest/terminal_statements.md

```diff
@@ -570,6 +570,65 @@ def f():
     reveal_type(x)  # revealed: Literal[1]
 ```
 
+## Calls to functions returning `Never` / `NoReturn`
+
+### No implicit return
+
+If we see a call to a function returning `Never`, we should be able to understand that the function
+cannot implicitly return `None`. In the below examples, verify that there are no errors emitted for
+invalid return type.
+
+```py
+from typing import NoReturn
+import sys
+
+def f() -> NoReturn:
+    sys.exit(1)
+```
+
+Let's try cases where the function annotated with `NoReturn` is some sub-expression.
+
+```py
+from typing import NoReturn
+import sys
+
+def _() -> NoReturn:
+    3 + sys.exit(1)
+
+def _() -> NoReturn:
+    3 if sys.exit(1) else 4
+```
+
+### Type narrowing
+
+```py
+from typing import NoReturn
+import sys
+
+def g(x: int | None):
+    if x is None:
+        sys.exit(1)
+
+    # TODO: should be just int, not int | None
+    reveal_type(x)  # revealed: int | None
```

Comment:
This does not work because of https://github.com/astral-sh/ty/issues/685 which I filed recently. I think that would need more discussion and better fixed in a follow up PR. I will _unmark_ this PR from fixing https://github.com/astral-sh/ty/issues/180, lest that main issue get closed automatically.

Reply:
I think it's still fine to close https://github.com/astral-sh/ty/issues/180 with this PR, as the general issue is fixed. We will still have https://github.com/astral-sh/ty/issues/685 open to track the particular issue (which isn't specific to calls returning `Never`).